src/resources/loader.js

```diff
@@ -73,24 +75,63 @@ Object.assign(pc, function () {
+                    if (err) {
+                        console.error(err);
+                        if (this._requests[key]) {
+                            for (var i = 0, len = this._requests[key].length; i < len; i++) {
+                                this._requests[key][i](err);
+                            }
+                        }
+                        delete this._requests[key];
                         return;
+                    }
 
-                    var i, len = this._requests[key].length;
-                    if (!err) {
-                        var resource = handler.open(url, data, asset);
-                        this._cache[key] = resource;
-                        for (i = 0; i < len; i++)
-                            this._requests[key][i](null, resource, extra);
-                    } else {
-                        for (i = 0; i < len; i++)
-                            this._requests[key][i](err);
+                    handler.load(urlObj, function (err, data, extra) {
+                        // make sure key exists because loader
+                        // might have been destroyed by now
+                        if (!this._requests[key])
+                            return;
+
+                        var i, len = this._requests[key].length;
+
+                        var resource;
+                        if (! err) {
+                            try {
+                                resource = handler.open(urlObj.original, data, asset);
+                            } catch (ex) {
+                                err = ex;
+                            }
+                        }
+
+                        if (!err) {
+                            this._cache[key] = resource;
+                            for (i = 0; i < len; i++)
+                                this._requests[key][i](null, resource, extra);
+                        } else {
+                            console.error(err);
+                            for (i = 0; i < len; i++)
+                                this._requests[key][i](err);
+                        }
+                        delete this._requests[key];
+                    }.bind(this), asset);
+                }.bind(this);
+
+                var normalizedUrl = url.split('?')[0];
+                if (this._app._enableBundles && this._app.bundles.hasUrl(normalizedUrl)) {
+                    if (!this._app.bundles.canLoadUrl(normalizedUrl)) {
+                        handleLoad('Bundle for ' + url + ' not loaded yet');
+                        return;
                     }
-                    delete this._requests[key];
-                }.bind(this), asset);
+
+                    this._app.bundles.loadUrl(normalizedUrl, function (err, fileUrlFromBundle) {
+                        handleLoad(err, { load: fileUrlFromBundle, original: url });
+                    });
+                } else {
+                    handleLoad(null, { load: url, original: url });
```

Comment:
Maybe we should define an actual class for this UrlObject? It's a bit confusing.